qiskit_experiments/library/tomography/basis/local_basis.py

```diff
@@ -0,0 +1,413 @@
+            and self._qubit_dim == getattr(value, "_qubit_dim", None)
+            and self._qubit_num_outcomes == getattr(value, "_qubit_num_outcomes", None)
+            and self._qubits == getattr(value, "_qubits", None)
+            and self._instructions == getattr(value, "_instructions", None)
+        )
+
+    def index_shape(self, qubits: Sequence[int]) -> Tuple[int]:
+        return len(qubits) * (self._size,)
+
+    def matrix_shape(self, qubits: Sequence[int]) -> Tuple[int]:
+        try:
+            return tuple(self._qubit_dim.get(i, self._default_dim) for i in qubits)
+        except TypeError as ex:
+            raise QiskitError("Invalid qubits for basis") from ex
+
+    def outcome_shape(self, qubits: Sequence[int]) -> Tuple[int]:
+        return tuple(self._qubit_num_outcomes.get(i, self._default_num_outcomes) for i in qubits)
+
+    def circuit(self, index: Sequence[int], qubits: Sequence[int]):
+        # pylint: disable = unused-argument
+        if not self._instructions:
+            raise NotImplementedError("This basis does not define circuits")
+        circuit = _tensor_product_circuit(self._instructions, index, self._name)
+        circuit.measure_all()
+        return circuit
+
+    def matrix(self, index: Sequence[int], outcome: int, qubits: Sequence[int]):
+        outcome_index = self._outcome_indices(outcome, tuple(qubits))
+        mat = 1
+        for idx, odx, qubit in zip(index, outcome_index, qubits):
+            povms = self._qubit_povms.get(qubit, self._default_povms)
+            mat = np.kron(povms[idx][odx], mat)
+        return mat
+
+    @functools.lru_cache(None)
+    def _outcome_indices(self, outcome: int, qubits: Tuple[int]) -> Tuple[int]:
+        """Convert an outcome integer to a tuple of single-qubit outcomes"""
+        num_outcomes = self._qubit_num_outcomes.get(qubits[0], self._default_num_outcomes)
+        try:
+            value = (outcome % num_outcomes,)
+            if len(qubits) == 1:
+                return value
+            return value + self._outcome_indices(outcome // num_outcomes, qubits[1:])
+        except TypeError as ex:
+            raise QiskitError("Invalid qubits for basis") from ex
+
+    def __json_encode__(self):
+        value = {
+            "name": self._name,
+            "instructions": list(self._instructions) if self._instructions else None,
+        }
+        if self._custom_defaults:
+            value["default_povms"] = list(self._default_povms)
+        if self._qubit_povms:
+            value["qubit_povms"] = self._qubit_povms
+        return value
+
+
+def _tensor_product_circuit(
+    instructions: Sequence[Instruction],
+    index: Sequence[int],
+    name: str = "",
+) -> QuantumCircuit:
+    """Return tensor product of 1-qubit basis instructions"""
+    size = len(instructions)
+    circuit = QuantumCircuit(len(index), name=f"{name}{list(index)}")
+    for i, elt in enumerate(index):
+        if elt >= size:
```

Comment:
This is actually checking `all(elt < len(instructions) for elt in index)` so not really anymore efficient to do outside since you need to iterate again.